### PR TITLE
Merge Relayer fee implementation into ConnextHandler

### PR DESCRIPTION
**Pull request by patitonar**
## The Problem
Subtask of https://github.com/connext/nxtp/issues/982

## The Solution
Merged the Relayer fee functionality implemented in the branch `feature/relayer-fee` for https://github.com/connext/nxtp/issues/900 into the new ConnextHandler structure

The merge included these commits:
- https://github.com/connext/nxtp/commit/ee7ab9e0ad05976fbc241921bbb4161e6cf733d4
- https://github.com/connext/nxtp/commit/05d84eef671d8694122dcfc6e103fb4b9a6286c4
- https://github.com/connext/nxtp/commit/ec7ca74ab4697b9adbff3d370b623a4012738bca
- https://github.com/connext/nxtp/commit/61af42135a5863e26e11f16507329343d2beebf9

## Checklist

- [ ] Test manually on `test-ui` using remote chains.
- [ ] Run load tests.
- [ ] Update documentation if needed.
- [ ] Update CHANGELOG.md.

### CODE REVIEW

File: packages/deployments/contracts/deploy/02_deployConnext.ts [hunk 1]
```diff
@@ -84,11 +84,13 @@ const func: DeployFunction = async (hre: HardhatRuntimeEnvironment): Promise<voi
       execute: {
         init: {
           methodName: "initialize",
+          // TODO - Use real RelayerFeeRouter
```
Comment (LayneHaber) [inline packages/deployments/contracts/deploy/02_deployConnext.ts:87]: will this go in a separate PR?

File: packages/deployments/contracts/test/connext.spec.ts [hunk 2]
```diff
@@ -132,11 +137,30 @@ describe("Connext", () => {
     const connextUtils = await deployContract("ConnextUtils");
     const routerPermissionsManagerLogic = await deployContract("RouterPermissionsManagerLogic");
 
+    // Deploy RelayerFeeRouters
+    originRelayerFeeRouter = await deployUpgradeableProxy<RelayerFeeRouter>(
+      "RelayerFeeRouter",
+      [originXappConnectionManager.address],
+      upgradeBeaconController.address,
+    );
```
Comment (LayneHaber) [inline packages/deployments/contracts/test/connext.spec.ts:145]: we dont have to use the same proxy layout as nomad -- we can use the ones that exist in the current `Connext.sol` so we can make use of the existing tools.

File: packages/deployments/contracts/test/connext.spec.ts [hunk 3]
```diff
@@ -132,11 +137,30 @@ describe("Connext", () => {
...
+
+    destinationRelayerFeeRouter = await deployUpgradeableProxy<RelayerFeeRouter>(
```
Comment (LayneHaber) [inline packages/deployments/contracts/test/connext.spec.ts:147]: see above

File: packages/deployments/contracts/test/connext.spec.ts [hunk 4]
```diff
@@ -132,11 +137,30 @@ describe("Connext", () => {
...
+      [destinationXappConnectionManager.address],
+      upgradeBeaconController.address,
+    );
+
     // Deploy bridge
     originBridge = (
       await deployUpgradeableProxy<ConnextHandler>(
         "ConnextHandler",
-        [originDomain, originXappConnectionManager.address, originTokenRegistry.address, weth.address],
+        [
```
Comment (LayneHaber) [inline packages/deployments/contracts/test/connext.spec.ts:157]: same refactor note as above (all of these can be made into separate issues)

File: packages/deployments/contracts/test/connext.spec.ts [hunk 5]
```diff
@@ -1107,4 +1198,109 @@ describe("Connext", () => {
       ).to.revertedWith("ConnextHandler__execute_maxRoutersExceeded()");
     });
   });
+
+  describe("relayerFee", () => {
```
Comment (LayneHaber) [inline packages/deployments/contracts/test/connext.spec.ts:1202]: we can leave these since they are already written but ultimately the tests should be written in foundry

**Review by LayneHaber — CHANGES_REQUESTED** [review]
The major change that should be implemented before merging is the change in the `XCalled` event signature

File: packages/deployments/contracts/contracts/nomad-xapps/contracts/connext/ConnextHandler.sol [hunk 6]
```diff
@@ -446,23 +493,37 @@ contract ConnextHandler is
...
+      // emit event
+      emit XCalled(
+        _transferId,
+        _args.params.to,
+        _args.params,
+        _transactingAssetId, // NOTE: this will switch from input to wrapper if native used
+        _bridged,
+        _amount,
+        _bridgedAmt,
+        0, // TODO use _args.relayerFee (stack too deep error)
```
[the remark this answers is not in the document]
Comment (patitonar) [reply packages/deployments/contracts/contracts/nomad-xapps/contracts/connext/ConnextHandler.sol:505]: This will be fixed in https://github.com/connext/nxtp/pull/1015

File: packages/deployments/contracts/contracts_forge/ConnextHandler.t.sol [hunk 7]
```diff
@@ -16,56 +21,145 @@ import "@openzeppelin/contracts/proxy/ERC1967/ERC1967Proxy.sol";
...
   // ============ Libraries ============
   using stdStorage for StdStorage;
 
   event MaxRoutersPerTransferUpdated(uint256 maxRouters, address caller);
+  event InitiatedClaim(uint32 indexed domain, address indexed recipient, address caller, bytes32[] transferIds);
+  event Claimed(address indexed recipient, uint256 total, bytes32[] transferIds);
+  event XCalled(
+    bytes32 indexed transferId,
+    address indexed to,
+    IConnext.CallParams params,
```
[the remark this answers is not in the document]
Comment (patitonar) [reply packages/deployments/contracts/contracts_forge/ConnextHandler.t.sol:44]: This will be fixed in https://github.com/connext/nxtp/pull/1015

File: packages/deployments/contracts/deploy/02_deployConnext.ts (hunk 1 above)
Comment (patitonar) [reply packages/deployments/contracts/deploy/02_deployConnext.ts:87]: This was already fixed in https://github.com/connext/nxtp/pull/1015

File: packages/deployments/contracts/contracts/nomad-xapps/contracts/relayer-fee-router/RelayerFeeRouter.sol [hunk 8]
```diff
@@ -0,0 +1,170 @@
...
+    bytes32 _sender,
+    bytes memory _message
+  ) external override onlyReplica onlyRemoteRouter(_origin, _sender) {
+    // parse recipient and transferIds from message
+    bytes29 _msg = _message.ref(0).mustBeClaimFees();
+
+    address recipient = _msg.recipient();
+    bytes32[] memory transferIds = _msg.transferIds();
+
+    // TODO - claim fee in connext
```
[the remark this answers is not in the document]
Comment (patitonar) [reply packages/deployments/contracts/contracts/nomad-xapps/contracts/relayer-fee-router/RelayerFeeRouter.sol]: Yes, done here https://github.com/connext/nxtp/pull/1026/commits/f186dce874db691d2fa7aa14342f46fa6e9a60b8

File: packages/deployments/contracts/test/connext.spec.ts (hunk 2 above)
Comment (patitonar) [reply packages/deployments/contracts/test/connext.spec.ts]: Updated the tests to use a transparent proxy for ConnextHandler and RelayerFeeRouter https://github.com/connext/nxtp/pull/1026/commits/c5361bfc30b6f4541676f8d3988be56dcd10c58e

File: packages/deployments/contracts/contracts/nomad-xapps/contracts/connext/ConnextHandler.sol [hunk 9]
```diff
@@ -403,6 +440,8 @@ contract ConnextHandler is
       revert ConnextHandler__xcall_emptyTo();
     }
 
+    if (_args.relayerFee == 0) revert ConnextHandler__xcall_relayerFeeIsZero();
```
[the remark this answers is not in the document]
Comment (patitonar) [reply packages/deployments/contracts/contracts/nomad-xapps/contracts/connext/ConnextHandler.sol]: Updated the logic here https://github.com/connext/nxtp/pull/1026/commits/d93ad94de59fb360e97f3727c7f80ba6d1c6a6b3

**Review by LayneHaber — APPROVED** [review]
